kola/switch-kernel: add a test for switching between rt and default kernel

zonggen opened the pull request:
Adds a sub-command that uploads the RT Kernel RPMs to a machine and switch between kernels using `rpm-ostree`.  The logic used to switch between Default Kernel and RT Kernel is defined in  https://github.com/openshift/machine-config-operator/blob/f363c7be6d2d506d900e196fa2e2d05ca08b93b6/pkg/daemon/update.go#L651.

Closes: https://issues.redhat.com/browse/GRPA-1439
Signed-off-by: Allen Bai <abai@redhat.com>

zonggen commented on mantle/kola/tests/rhcos/switch-kernel.go:
```diff
@@ -0,0 +1,167 @@
+
+func testSwitchKernel(c cluster.TestCluster) {
+	m := c.Machines()[0]
+	kernelRtRpmDir := "/srv/kernel-rt/"
```
Thanks for the feedback! I wanted to change this line to read from an option `--rpm-dir` or similar but that also means we need to set a default value for it, which goes back to hard-coding path. Making it a sub-command seems like the way to go.

zonggen commented:
cc @arithx 
Could you take a second look to see it's reasonable?

zonggen commented on mantle/cmd/kola/switchkernel.go:
```diff
@@ -0,0 +1,249 @@
+)
+
+var (
+	homeDir            = `/var/home/core`
```
Can I assume `core` user is always available in a `platform.Machine`?

cgwalters replied:
Yeah.

cgwalters commented on mantle/cmd/kola/switchkernel.go:
```diff
@@ -0,0 +1,249 @@
+
+	err := dropRpmFilesAll(m, rtKernelRpmDir)
+	if err != nil {
+		return fmt.Errorf("Failed dropping Kernel RT RPM files: %v", err)
```
In new code let's use `errors.Wrapf()`.

zonggen commented:
<details>
<summary>sample run & output</summary>

```
[coreos-assembler]$ /coreos-assembler/mantle/bin/kola switch-kernel --kernel-rt ./kernel-rt -b rhcos --ignition-version v2 --qemu-image rhcos-44.81.202003021844-0-qemu.x86_64.qcow2 
Dropping RT Kernel RPMs...
Switching from Default to RT Kernel...
+ FROM_KERNEL=default
+ TO_KERNEL=rt-kernel
+ DEFAULT_KERNEL_PKG='kernel kernel-core kernel-modules kernel-modules-extra'
+ RT_KERNEL_PKG='kernel-rt-core kernel-rt-modules kernel-rt-modules-extra'
+ [[ default == \d\e\f\a\u\l\t ]]
+ [[ rt-kernel == \r\t\-\k\e\r\n\e\l ]]
+ RT_KERNEL_REPO=/var/home/core/kernel-rt-rpms/
++ ls /var/home/core/kernel-rt-rpms/
+ [[ -z kernel-rt-core-4.18.0-147.5.1.rt24.98.el8_1.x86_64.rpm
kernel-rt-modules-4.18.0-147.5.1.rt24.98.el8_1.x86_64.rpm
kernel-rt-modules-extra-4.18.0-147.5.1.rt24.98.el8_1.x86_64.rpm ]]
+ ARGS='override remove kernel kernel-core kernel-modules kernel-modules-extra'
++ ls /var/home/core/kernel-rt-rpms/
+ for RPM in $(ls ${RT_KERNEL_REPO})
+ ARGS+=' --install /var/home/core/kernel-rt-rpms//kernel-rt-core-4.18.0-147.5.1.rt24.98.el8_1.x86_64.rpm'
+ for RPM in $(ls ${RT_KERNEL_REPO})
+ ARGS+=' --install /var/home/core/kernel-rt-rpms//kernel-rt-modules-4.18.0-147.5.1.rt24.98.el8_1.x86_64.rpm'
+ for RPM in $(ls ${RT_KERNEL_REPO})
+ ARGS+=' --install /var/home/core/kernel-rt-rpms//kernel-rt-modules-extra-4.18.0-147.5.1.rt24.98.el8_1.x86_64.rpm'
+ rpm-ostree override remove kernel kernel-core kernel-modules kernel-modules-extra --install /var/home/core/kernel-rt-rpms//kernel-rt-core-4.18.0-147.5.1.rt24.98.el8_1.x86_64.rpm --install /var/home/core/kernel-rt-rpms//kernel-rt-modules-4.18.0-147.5.1.rt24.98.el8_1.x86_64.rpm --install /var/home/core/kernel-rt-rpms//kernel-rt-modules-extra-4.18.0-147.5.1.rt24.98.el8_1.x86_64.rpm
Checking out tree a35c3a6...done
Enabled rpm-md repositories: rhel8-baseos rhel8-appstream rhel8-rt
Updating metadata for 'rhel8-baseos'...done
rpm-md repo 'rhel8-baseos'; generated: 2020-02-27T15:31:54Z
Updating metadata for 'rhel8-appstream'...done
rpm-md repo 'rhel8-appstream'; generated: 2020-03-04T17:20:00Z
Updating metadata for 'rhel8-rt'...done
rpm-md repo 'rhel8-rt'; generated: 2020-02-25T05:36:45Z
Importing rpm-md...done
Resolving dependencies...done
Applying 4 overrides and 3 overlays
Processing packages...done
Running pre scripts...done
Running post scripts...done
Running posttrans scripts...done
Writing rpmdb...done
Generating initramfs...done
Writing OSTree commit...done
Staging deployment...done
Removed:
  kernel-4.18.0-147.5.1.el8_1.x86_64
  kernel-core-4.18.0-147.5.1.el8_1.x86_64
  kernel-modules-4.18.0-147.5.1.el8_1.x86_64
  kernel-modules-extra-4.18.0-147.5.1.el8_1.x86_64
Added:
  kernel-rt-core-4.18.0-147.5.1.rt24.98.el8_1.x86_64
  kernel-rt-modules-4.18.0-147.5.1.rt24.98.el8_1.x86_64
  kernel-rt-modules-extra-4.18.0-147.5.1.rt24.98.el8_1.x86_64
Run "systemctl reboot" to start a reboot
Rebooting machine...
Checking kernel type...
Switched to RT Kernel successfully!
Switching from RT to Default Kernel...
+ FROM_KERNEL=rt-kernel
+ TO_KERNEL=default
+ DEFAULT_KERNEL_PKG='kernel kernel-core kernel-modules kernel-modules-extra'
+ RT_KERNEL_PKG='kernel-rt-core kernel-rt-modules kernel-rt-modules-extra'
+ [[ rt-kernel == \d\e\f\a\u\l\t ]]
+ [[ rt-kernel == \r\t\-\k\e\r\n\e\l ]]
+ [[ default == \d\e\f\a\u\l\t ]]
+ ARGS='override reset kernel kernel-core kernel-modules kernel-modules-extra'
+ for PKG in $RT_KERNEL_PKG
+ ARGS+=' --uninstall kernel-rt-core'
+ for PKG in $RT_KERNEL_PKG
+ ARGS+=' --uninstall kernel-rt-modules'
+ for PKG in $RT_KERNEL_PKG
+ ARGS+=' --uninstall kernel-rt-modules-extra'
+ rpm-ostree override reset kernel kernel-core kernel-modules kernel-modules-extra --uninstall kernel-rt-core --uninstall kernel-rt-modules --uninstall kernel-rt-modules-extra
Staging deployment...done
Removed:
  kernel-rt-core-4.18.0-147.5.1.rt24.98.el8_1.x86_64
  kernel-rt-modules-4.18.0-147.5.1.rt24.98.el8_1.x86_64
  kernel-rt-modules-extra-4.18.0-147.5.1.rt24.98.el8_1.x86_64
Added:
  kernel-4.18.0-147.5.1.el8_1.x86_64
  kernel-core-4.18.0-147.5.1.el8_1.x86_64
  kernel-modules-4.18.0-147.5.1.el8_1.x86_64
  kernel-modules-extra-4.18.0-147.5.1.el8_1.x86_64
Run "systemctl reboot" to start a reboot
Rebooting machine...
Checking kernel type...
Switched back to Default Kernel successfully!
```
</details>

cgwalters commented on mantle/cmd/kola/switchkernel.go:
```diff
@@ -0,0 +1,249 @@
+		}
+	}`, base64.StdEncoding.EncodeToString([]byte(switchKernelScript))))
+
+	flight, err := kola.NewFlight("qemu-unpriv")
```
There's nothing intrinsically specific to qemu about this, right?  I think you can use `kola.NewFlight(kolaPlatform)` and drop the `kolaPlatform` check above.

zonggen replied:
Ah, yes that makes sense.

arithx commented:
Seems reasonable to me; I agree with @cgwalters' comments but after those are fixed should be fine.

zonggen commented:
Switched to `errors.Wrapf` and de-hardcoded `qemu-unpriv`.

Rebased and squashed into one commit.

cgwalters commented on mantle/cmd/kola/switchkernel.go:
```diff
@@ -0,0 +1,249 @@
+
+	flight, err := kola.NewFlight(kolaPlatform)
+	if err != nil {
+		return errors.Wrapf(err, "Flight failed: %v", err)
```
The nice thing about `errors.Wrapf` is you *don't* need to do the `: %v", err)`; it does that automatically.  And it also preserves the underlying error so it's not turned into a string.

cgwalters replied:
(Search for other places in the code that use it)

zonggen commented:
Fixed `: %v", err` in `errors.Wrapf` and update the error message to be consistent with other error message.

cgwalters commented on mantle/cmd/kola/switchkernel.go:
```diff
@@ -0,0 +1,249 @@
+	fmt.Println("Checking kernel type...")
+	cmd = "uname -v | grep -q 'PREEMPT RT'"
+	_, _, err = m.SSH(cmd)
+	if err == nil {
```
This should be `err != nil` right?

Hmm...I think this test is passing because you're missing a `_` in `PREEMPT_RT` so both tests are inverted?

zonggen replied:
Double checked inside a rhcos machine..
```
[core@ibm-p8-kvm-03-guest-02 ~]$ uname -v 
#1 SMP PREEMPT RT Tue Jan 14 16:03:46 UTC 2020
```

cgwalters replied:
Ah OK.  But then I think we should invert these checks, because `if err == nil` is a HUGE trap in Go, it just looks like a typo.

And specifically here the `error.Wrapf()` is wrong for this one because `err` is nil, so it will return `nil`, and the rest of the code will think no error occurred.

zonggen replied:
The `make check` ran by `openshift-ci-rebot`  didn't run `switchkernel.go` at all, it only ran `kola --help` for sanity check.

Yes, the it should be `if err != nil`, sorry should've mentioned earlier.. and I confirmed on RHCOS the original test did not produce expected result

cgwalters commented on mantle/cmd/kola/switchkernel.go:
```diff
@@ -0,0 +1,249 @@
+
+	// check if the kernel has switched back to default kernel
+	fmt.Println("Checking kernel type...")
+	cmd = "! $(uname -v | grep -q 'PREEMPT RT')"
```
I think it's better to use `grep -v` then `! grep` - it more clearly expresses intent (and ensures that you still *do* get an error if e.g. an input file doesn't exist or grep got OOM killed or whatever).

And see above too about the `_`.

zonggen replied:
Updated with `grep -v` and tested on a RHCOS machine, worked without issue

zonggen commented on mantle/cmd/kola/switchkernel.go:
```diff
@@ -0,0 +1,249 @@
+	cmd := "sudo " + homeDir + "/switch-kernel.sh rt-kernel default"
+	stdout, stderr, err := m.SSH(cmd)
+	if err != nil {
+		return errors.Wrapf(err, "failed to run %", cmd)
```
cmd/kola/switchkernel.go:206:10: Wrapf format % is missing verb at end of string

arithx replied:
You're missing the `s` of the `%s` in the format string.

zonggen replied:
Yep I wanted to make a note I missed that

zonggen commented:
Updated and tested locally

<details>
<summary>Output</summary>
```
[coreos-assembler]$ /coreos-assembler/mantle/bin/kola switch-kernel -b rhcos --ignition-version v2 --qemu-image rhcos-44.81.202003021844-0-qemu.x86_64.qcow2 --kernel-rt kernel-/
Dropping RT Kernel RPMs...
Switching from Default to RT Kernel...
+ FROM_KERNEL=default
+ TO_KERNEL=rt-kernel
+ DEFAULT_KERNEL_PKG='kernel kernel-core kernel-modules kernel-modules-extra'
+ RT_KERNEL_PKG='kernel-rt-core kernel-rt-modules kernel-rt-modules-extra'
+ [[ default == \d\e\f\a\u\l\t ]]
+ [[ rt-kernel == \r\t\-\k\e\r\n\e\l ]]
+ RT_KERNEL_REPO=/var/home/core/kernel-rt-rpms/
++ ls /var/home/core/kernel-rt-rpms/
+ [[ -z kernel-rt-core-4.18.0-147.5.1.rt24.98.el8_1.x86_64.rpm
kernel-rt-modules-4.18.0-147.5.1.rt24.98.el8_1.x86_64.rpm
kernel-rt-modules-extra-4.18.0-147.5.1.rt24.98.el8_1.x86_64.rpm ]]
+ ARGS='override remove kernel kernel-core kernel-modules kernel-modules-extra'
++ ls /var/home/core/kernel-rt-rpms/
+ for RPM in $(ls ${RT_KERNEL_REPO})
+ ARGS+=' --install /var/home/core/kernel-rt-rpms//kernel-rt-core-4.18.0-147.5.1.rt24.98.el8_1.x86_64.rpm'
+ for RPM in $(ls ${RT_KERNEL_REPO})
+ ARGS+=' --install /var/home/core/kernel-rt-rpms//kernel-rt-modules-4.18.0-147.5.1.rt24.98.el8_1.x86_64.rpm'
+ for RPM in $(ls ${RT_KERNEL_REPO})
+ ARGS+=' --install /var/home/core/kernel-rt-rpms//kernel-rt-modules-extra-4.18.0-147.5.1.rt24.98.el8_1.x86_64.rpm'
+ rpm-ostree override remove kernel kernel-core kernel-modules kernel-modules-extra --install /var/home/core/kernel-rt-rpms//kernel-rt-core-4.18.0-147.5.1.rt24.98.el8_1.x86_64.m
Checking out tree a35c3a6...done
Enabled rpm-md repositories: rhel8-baseos rhel8-appstream rhel8-rt
Updating metadata for 'rhel8-baseos'...done
rpm-md repo 'rhel8-baseos'; generated: 2020-02-27T15:31:54Z
Updating metadata for 'rhel8-appstream'...done
rpm-md repo 'rhel8-appstream'; generated: 2020-03-04T17:20:00Z
Updating metadata for 'rhel8-rt'...done
rpm-md repo 'rhel8-rt'; generated: 2020-02-25T05:36:45Z
Importing rpm-md...done
Resolving dependencies...done
Applying 4 overrides and 3 overlays
Processing packages...done
Running pre scripts...done
Running post scripts...done
Running posttrans scripts...done
Writing rpmdb...done
Generating initramfs...done
Writing OSTree commit...done
Staging deployment...done
Removed:
  kernel-4.18.0-147.5.1.el8_1.x86_64
  kernel-core-4.18.0-147.5.1.el8_1.x86_64
  kernel-modules-4.18.0-147.5.1.el8_1.x86_64
  kernel-modules-extra-4.18.0-147.5.1.el8_1.x86_64
Added:
  kernel-rt-core-4.18.0-147.5.1.rt24.98.el8_1.x86_64
  kernel-rt-modules-4.18.0-147.5.1.rt24.98.el8_1.x86_64
  kernel-rt-modules-extra-4.18.0-147.5.1.rt24.98.el8_1.x86_64
Run "systemctl reboot" to start a reboot
Rebooting machine...
Checking kernel type...
Switched to RT Kernel successfully!
Switching from RT to Default Kernel...
+ FROM_KERNEL=rt-kernel
+ TO_KERNEL=default
+ DEFAULT_KERNEL_PKG='kernel kernel-core kernel-modules kernel-modules-extra'
+ RT_KERNEL_PKG='kernel-rt-core kernel-rt-modules kernel-rt-modules-extra'
+ [[ rt-kernel == \d\e\f\a\u\l\t ]]
+ [[ rt-kernel == \r\t\-\k\e\r\n\e\l ]]
+ [[ default == \d\e\f\a\u\l\t ]]
+ ARGS='override reset kernel kernel-core kernel-modules kernel-modules-extra'
+ for PKG in $RT_KERNEL_PKG
+ ARGS+=' --uninstall kernel-rt-core'
+ for PKG in $RT_KERNEL_PKG
+ ARGS+=' --uninstall kernel-rt-modules'
+ for PKG in $RT_KERNEL_PKG
+ ARGS+=' --uninstall kernel-rt-modules-extra'
+ rpm-ostree override reset kernel kernel-core kernel-modules kernel-modules-extra --uninstall kernel-rt-core --uninstall kernel-rt-modules --uninstall kernel-rt-modules-extra
Staging deployment...done
Removed:
  kernel-rt-core-4.18.0-147.5.1.rt24.98.el8_1.x86_64
  kernel-rt-modules-4.18.0-147.5.1.rt24.98.el8_1.x86_64
  kernel-rt-modules-extra-4.18.0-147.5.1.rt24.98.el8_1.x86_64
Added:
  kernel-4.18.0-147.5.1.el8_1.x86_64
  kernel-core-4.18.0-147.5.1.el8_1.x86_64
  kernel-modules-4.18.0-147.5.1.el8_1.x86_64
  kernel-modules-extra-4.18.0-147.5.1.el8_1.x86_64
Run "systemctl reboot" to start a reboot
Rebooting machine...
Checking kernel type...
Switched back to Default Kernel successfully!

```
</details>

cgwalters commented:
/lgtm

zonggen commented:
Thank you @cgwalters, merging

cgwalters commented:
> Thank you @cgwalters, merging

Currently this repo uses Prow, so the bot will merge.
